ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,391 @@
+   * @param capacityInBytes the intended capacity of the segment
+   * @param metrics the {@link StoreMetrics} instance to use.
+   * @throws IOException if the file cannot be read or created
+   */
+  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics)
+      throws IOException {
+    if (!file.exists() || !file.isFile()) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + " does not exist or is not a file");
+    }
+    this.file = file;
+    this.name = name;
+    this.capacityInBytes = capacityInBytes;
+    this.metrics = metrics;
+    fileChannel = Utils.openChannel(file, true);
+    segmentView = new Pair<>(file, fileChannel);
+    // state is always initialized to FREE and the end offset is always initialized to 0.
+    // externals will set the correct values for these two variables
+    state = State.FREE;
+    endOffset = new AtomicLong(0);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Guarantees that the {@code buffer} is either written in its entirety in this segment or is transparently forwarded
+   * to the next segment (if one exists).
+   * <p/>
+   * Can cause state change from {@link State#FREE} -> {@link State#ACTIVE} if this is the first write to the segment.
+   * Can cause state change from {@link State#ACTIVE} -> {@link State#SEALED} if this write cannot be accommodated in
+   * the segment.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if there is not enough space for {@code buffer} and forwarding is not possible
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
+      throws IOException {
+    int bytesWritten = 0;
+    if (state.equals(State.SEALED)) {
+      ensureNext(metrics.overflowWriteError);
+      bytesWritten = next.appendFrom(buffer);
+    } else if (endOffset.get() + buffer.remaining() > capacityInBytes) {
+      // maintain the invariant that there are no "partial" writes. A log segment should write the data that it receives
+      // completely and the data that was provided in a single call cannot be written across multiple segments.
+      ensureNext(metrics.overflowWriteError);
+      logger.info("{} : Rolling over writes to {} on write of data of size {} from a buffer. "
+              + "Current end offset is {} and capacity is {}", file.getAbsolutePath(), next.getName(),
+          buffer.remaining(), getEndOffset(), capacityInBytes);
+      state = State.SEALED;
+      bytesWritten = next.appendFrom(buffer);
+    } else {
+      state = State.ACTIVE;
+      while (buffer.hasRemaining()) {
+        bytesWritten += fileChannel.write(buffer, endOffset.get());
+      }
+      endOffset.addAndGet(bytesWritten);
+    }
+    return bytesWritten;
+  }
```

Comment:
The `LogSegment` `next` determines ordering. Nobody else (needs to) stores the ordering.